File: C++/TopKFrequent.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int, int> m;
        for (auto num : nums) {
            m[num]++;
        }

        vector<vector<int>> buckets(n+1);

        for (auto pair : m) {
            buckets[pair.second].push_back(pair.first);
        }

        vector<int> ans;

        for (int i = n; i > 0; i--) {
            if (ans.size() >= k) {
                break;
            }
            if (!buckets[i].empty()) {
                ans.insert(ans.end(), buckets[i].begin(), buckets[i].end());
            }
        }
        return ans;
    }
};

File: Python/top_k_frequent.py
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        num_freq = {}
        for num in nums:
            num_freq[num] = 1 + num_freq.get(num, 0)
        lst = [[] for i in range(len(nums) + 1)]
        for key, value in num_freq.items():
            lst[value].append(key)
        ans = []
        for i in range(len(lst) - 1, 0, -1):
            ans.extend(lst[i])
            if len(ans) >= k:
                return ans

# sol = Solution()
# nums = [1,1,1,2,2,3]
# print(sol.topKFrequent(nums, 2))
